### ISSUE TITLE
Bug - Source connector made with connector Builder do not recognize the last date modification cursor in the connection state "streamState" in incrementa sync

### ISSUE THREAD

**Issue by leonmm2** [airbytehq/airbyte#61567]
### Topic

Connector Builder bug in incremental sync conections

### Relevant information

Hello community.

I have developed a source connector for Dolibarr ERP in the Connector Builder UI, in this connector I configured several parent streams that support standar pagination and incremental sync for an API with No Filter by date capability, also I configured some child streams that depends on the parent streams and in this child streams I enable the switch indicating that their paren stream have incremental sync and validated that the parent stream cursor is updated when the child stream data is modified or updated. 
For the incremental sync in the parent streams I configured the `sortfield` with the corresponding `date_modification` parameter and `sortorder` in descendent as required by Airbyte for incremental sync without date filtering. Also I configured the cursor field for the incremental sync with the corresponding date_modification variable for each parent stream.
The problem I identified is that in the first sync with this connector with the `Incremental | Append + Dedup` enabled in the parent streams the sincronization works well and the `streamState` for the connection is saved correctly with the last date modification of the data extracted but, when I make some changes in the origin data and try the second sync, the connector again extract all the records from the child streams and the first page of the parent streams and take so long time to do the extract, load and deduplication (much more than in the first sync) as if the connector where ignoring the last `streamState`. The AI agent in the slack community guided me in all verifications and the conclusion is a possible bug in the Connector Builder that generates a logic for the conector that ignores the last `streamState` for the incrementas sync in incremental parent streams with child streams.

I attach the YAML code in a txt file of the connector and the link to the PR of the Dolibarr connector i have developed:
https://github.com/airbytehq/airbyte/pull/61388

[dolibarr.txt](https://github.com/user-attachments/files/20720937/dolibarr.txt)

**Comment by leonmm2** [comment]
Hi Airbyte team,

Below is an analysis generated by GitHub Copilot (AI assistant) based on the YAML I provided for the declarative Dolibarr ERP connector, regarding the issue with incremental sync and the recognition of the streamState in parent-child stream connections.

---

### YAML Analysis (by GitHub Copilot)

- **Structure:**  
  The YAML is well-formed, following Airbyte's recommended structure for declarative sources. Each main and secondary stream is correctly defined, including schemas and transformations according to the documentation.

- **Incremental Sync:**  
  The main streams (e.g., Products list, Customers list, etc.) correctly use `DatetimeBasedCursor` for incremental sync. The `cursor_field` is set appropriately, and the date format is consistent.

- **Substreams and incremental dependency:**  
  Substreams (e.g., invoice lines and payments) use `partition_router` with `incremental_dependency: true`, which indicates they should rely on the incremental state of the parent stream, according to the official docs.

- **Additional configurations:**  
  - Date formats (`cursor_datetime_formats`) are properly defined.
  - The `start_datetime` parameter uses the provided config variable.
  - Transforms, paginators, and requesters are all correctly configured.

---

### Conclusion

- **The YAML correctly implements** Airbyte's best practices and recommendations for incremental sync with parent-child dependencies.
- The issue does not appear to be a configuration problem, but rather a behavior in the Connector Builder/Core: the `streamState` is saved but not respected during subsequent syncs, causing a full re-extraction from the substreams.
- This suggests it is likely a limitation or bug in the platform, not in the connector YAML itself.

If you need any specific YAML fragment or logs to help confirm this behavior, I can provide them.

Let me know if you need any further information.

---

**Comment by leonmm2** [comment]
### Investigation: State Propagation Bug in `SubstreamPartitionRouter` (Concurrent CDK)

I've been analyzing a severe performance degradation issue affecting connectors that leverage `incremental_dependency: true` with parent-child streams. After tracing the execution paths in the latest declarative CDK (specifically `airbyte_cdk/sources/declarative/partition_routers/substream_partition_router.py`) using GitHub Copilot to map the codebase, there is a structural state propagation gap causing child streams to lose their cursor context.

#### Problem Summary

When configuring a stream with `incremental_dependency: true`:
- **Initial Sync (Sync 1):** Works as expected; the `streamState` is persisted with the parent cursor checkpoints.
- **Subsequent Syncs (Sync 2+):** The connector restarts data extraction from the very beginning, re-extracting the entire dataset (making syncs 100x to 1000x slower).

The breakdown occurs because the state structure shifts from the expected nested wrapper to a flat object where the context is lost.

*Expected Structure:*
```json
{
  "parent_state": {
    "parent_stream_name": {"date_modification": "2024-08-01T00:00:00Z"}
  },
  "states": [...],
  "lookback_window": 0
}
```

*Actual Broken Structure:*
```json
{
  "date_modification": "2024-08-01T00:00:00Z"
}
```
Without the `parent_state` envelope, the child stream cannot isolate or resolve the parent cursor dependency during subsequent execution cycles within the new Concurrent CDK architecture.

---

#### Proposed Connector-Level Workaround

For developers currently blocked by this behavior (e.g., in connectors like `source-dolibarr` which relies heavily on `incremental_dependency` for its lines and payments streams), it is possible to bypass the core CDK limitation by implementing a custom state migration to repair the JSON structure before the CDK parses it, and a custom router to inject the state into the parent stream reader.

**1. Custom Components Implementation (`components.py`)**
Create a `components.py` file in your connector directory. Note that the cursors array includes `date_modification` and `tms` to match the specific parent streams being used:

```python
from dataclasses import dataclass
from typing import Any, Iterable, Mapping, Optional

from airbyte_cdk.sources.declarative.partition_routers.substream_partition_router import SubstreamPartitionRouter
from airbyte_cdk.sources.declarative.state_migration import StateMigration
from airbyte_cdk.sources.declarative.types import StreamSlice, StreamState

@dataclass
class IncrementalSyncStateMigration(StateMigration):
    """
    Transforms flattened state structures back into the required nested format
    containing the parent_state wrapper to preserve incremental_dependency.
    """
    def should_migrate(self, stream_state: Mapping[str, Any]) -> bool:
        return "parent_state" not in stream_state and stream_state and len(stream_state) > 0

    def migrate(self, stream_state: Mapping[str, Any]) -> Mapping[str, Any]:
        if not stream_state:
            return stream_state

        migrated_parent_state = {}
        for key, value in stream_state.items():
            if key != "states" and isinstance(value, dict) and any(
                cursor in value for cursor in ["date_modification", "tms", "updated_at", "timestamp", "created_at", "modified"]
            ):
                migrated_parent_state[key] = value

        return {
            "parent_state": migrated_parent_state,
            "states": [],
            "lookback_window": 0,
        }

@dataclass
class IncrementalDependencySubstreamPartitionRouter(SubstreamPartitionRouter):
    """
    Custom router that explicitly extracts the parent stream state 
    and injects it into the parent's read_records execution loop.
    """
    def stream_slices(self, stream_state: Optional[StreamState] = None) -> Iterable[StreamSlice]:
        if not self.parent_stream_configs:
            yield from []
            return

        for parent_stream_config in self.parent_stream_configs:
            parent_stream = parent_stream_config.stream
            parent_key = parent_stream_config.parent_key.eval(self.config)
            partition_field = parent_stream_config.partition_field.eval(self.config)

            parent_stream_state = None
            if stream_state and isinstance(stream_state, dict):
                state_dict = stream_state.get("parent_state", stream_state)
                parent_stream_state = state_dict.get(parent_stream.name)

            for parent_record in parent_stream.read_records(
                sync_mode=self.sync_mode,
                stream_state=parent_stream_state,
            ):
                cursor_slice = {}
                if parent_stream_state:
                    cursor_slice["parent_state"] = parent_stream_state
                
                if parent_stream_config.incremental_dependency:
                    cursor_slice["incremental_dependency"] = True
                    cursor_slice["parent_stream_name"] = parent_stream.name

                yield StreamSlice(
                    partition={partition_field: parent_record[parent_key]},
                    cursor_slice=cursor_slice,
                )
```

**2. Declarative Manifest Integration (`manifest.yaml`)**

First, register the state migration at the root of your `definitions` block:

```yaml
definitions:
  state_migration:
    type: IncrementalSyncStateMigration
  # ... other definitions ...
```

Then, for EVERY stream that has `incremental_dependency: true` (in Dolibarr this includes `Products category id list`, `Customer invoices lines list`, `Supplier invoices lines list`, `Customer invoices payments list`, `Supplier invoices payments list`), change the partition router `type` to use your custom class. 

Example for `Customer invoices lines list`:
```yaml
    Customer invoices lines list:
      type: DeclarativeStream
      # ...
      retriever:
        type: SimpleRetriever
        # ...
        partition_router:
          type: IncrementalDependencySubstreamPartitionRouter # <--- Changed from SubstreamPartitionRouter
          parent_stream_configs:
            - type: ParentStreamConfig
              parent_key: invoice_id
              partition_field: invoice_id
              stream:
                $ref: "#/definitions/streams/Customer invoices list"
              incremental_dependency: true
```

*(Continued in the next comment for the proposed CDK Core fix)*

**Comment by leonmm2** [comment]
#### Root Cause in Airbyte Core CDK

Reviewing the current implementation of `airbyte_cdk/sources/declarative/partition_routers/substream_partition_router.py`, the state propagation gap seems to originate in two specific areas handling the new concurrent stream logic:

**1. Hardcoded Empty Cursor Slice (Line ~214)**
Inside the `stream_slices` generator, when iterating over `iterate_with_last_flag(partition.read())`, the `StreamSlice` is yielded with a hardcoded empty `cursor_slice`:

```python
yield StreamSlice(
    partition={
        partition_field: partition_value,
        "parent_slice": parent_partition or {},
    },
    cursor_slice={},  # <-- Context lost here
    extra_fields=extracted_extra_fields,
)
```
Because the `cursor_slice` is empty, downstream components relying on the slice to carry the parent's checkpoint fail to associate the exact cursor bounds.

**2. Brittle State Migration (`_migrate_child_state_to_parent_state`)**
While the CDK attempts to mitigate this via `_migrate_child_state_to_parent_state()` (Line ~261), the logic assumes very specific structures (e.g., checking strictly for `{"state": ...}`) and fails to correctly reconstruct the nested `parent_state` envelope required by child streams when multiple cursors or complex dates are involved.

#### Affected Connectors Landscape

This architecture gap likely degrades performance on several production connectors that rely on `incremental_dependency`. For instance:
- `source-gmail`
- `source-mixpanel`
- `source-instagram`
- `source-typeform`

Interestingly, checking the source tree reveals that `source-intercom` and `source-jira` have already bypassed this by implementing similar custom `StateMigration` wrappers inside their local codebases, confirming that the CDK is currently dropping this context globally.

#### Local Testing and Verification

For anyone implementing the connector-level state migration workaround, you can confirm whether delta syncs are functioning properly by piping the states in your local terminal:

```bash
# Execute initial sync and capture the generated state object
STATE=$(docker-compose run connector read config.json catalog.json | grep -A1 '"type":"STATE"' | tail -1 | jq -r '.state')

# Run the subsequent sync cycle passing the saved state via standard input
docker-compose run connector read config.json catalog.json <<< "$STATE" > sync2_output.json

# Count the record types processed in the delta sync
grep '"type":"RECORD"' sync2_output.json | wc -l
```
*Success Metric:* The record count on the second run should be drastically lower, and the logs should indicate that the parent stream is reading from its saved cursor rather than from the beginning.

I'd welcome feedback from the maintainers. Does fixing the `cursor_slice` propagation and strengthening the `_migrate_child_state_to_parent_state` helper align with the roadmap for the concurrent CDK architecture?

**Comment by lazebnyi** [comment]
Hi @leonmm2, thanks for the detailed report and for attaching the manifest.

TL;DR: The behavior you observed is reproducible, but it is not caused by the Connector Builder generating broken logic or by the CDK ignoring saved state in general. The root cause is that your child streams (e.g. Products category id list, Customer invoices lines list) are full-refresh streams (they have no incremental_sync of their own) with incremental_dependency: true. In that configuration the parent's cursor is never persisted across sync jobs, so every sync re-reads the parents from the beginning and re-fetches all child partitions.

Why this happens:

1. The parent cursor checkpoint is stored inside the child stream's state, under a parent_state key. This only happens when the child stream itself is incremental (it then uses a per-partition cursor that carries parent_state). A full-refresh child stream only checkpoints an internal completion sentinel, no parent_state is ever written.
2. Even if it were written, the platform discards full-refresh stream state at the start of each new sync job (it is only kept for resuming within a job). So a full-refresh child has nothing to resume from, by design.

This also explains what you saw: your parent streams (synced as Incremental | Append + Dedup) respected their own saved state — that's why only the first page was re-read there. The full re-extraction happened only for the parents being read internally by the child streams' partition routers.

How to fix your connector: add incremental_sync to the child streams themselves, using a cursor field present on the child records (e.g. date_modification where the API returns it). Once the child stream is incremental, its state will include parent_state, and the second sync will read only parents modified since the last checkpoint (the data-feed stop condition will stop pagination early) and only re-fetch children of those parents. I verified this end-to-end against the latest CDK with a mock server modeled on your manifest: with an incremental child, sync 2 made a single parent page request and fetched only the new/updated partitions; with a full-refresh child it re-read everything.

For child streams whose records have no usable cursor field, the `incremental_dependency` toggle currently provides no benefit across sync jobs, you can remove it there.

cc @pnilan
